common/Model/ProductReview.php

```diff
@@ -0,0 +1,166 @@
+
+    private ReviewFactory $reviewFactory;
+
+    private RatingFactory $ratingFactory;
+
+    private CustomerRepositoryInterface $customerInterface;
+
+    private Registry $registry;
+
+    private RatingVoteCollection $ratingVoteCollection;
+
+    public function __construct(
+        InvitationHelper $invitationHelper,
+        LoggerInterface $logger,
+        ProductRepository $productRepository,
+        ReviewFactory $reviewFactory,
+        RatingFactory $ratingFactory,
+        CustomerRepositoryInterface $customerInterface,
+        Registry $registry,
+        RatingVoteCollection $ratingVoteCollection
+    ) {
+        $this->invitationHelper = $invitationHelper;
+        $this->logger = $logger;
+        $this->productRepository = $productRepository;
+        $this->reviewFactory = $reviewFactory;
+        $this->ratingFactory = $ratingFactory;
+        $this->customerInterface = $customerInterface;
+        $this->registry = $registry;
+        $this->ratingVoteCollection = $ratingVoteCollection;
+    }
+
+    public function sync(array $requestData): ?int {
+        $productReview = $requestData['product_review'];
+
+        try {
+            $product = $this->productRepository->getById($productReview['product_id']);
+        } catch (NoSuchEntityException $e) {
+            throw new NotFoundException(sprintf('Could not find product with ID (%d)', $productReview['product_id']));
+        }
+
+        $storeId = $product->getStoreId();
+        $config = $this->invitationHelper->getConfigData($storeId);
+        $this->isAuthorized($config, $requestData['webshop_id'], $requestData['api_key']);
+
+        if (!$config['product_reviews']) {
+            throw new ForbidenException('Product review sync is disabled');
+        }
+
+        if ($productReview['deleted']) {
+            $this->registry->register('isSecureArea', true);
+            $this->reviewFactory->create()->setId($productReview['id'])->delete();
+            return null;
+        }
+
+        $review = $this->reviewFactory->create()
+            ->setId($productReview['id'])
+            ->setEntityPkValue($productReview['product_id'])
+            ->setStatusId(Review::STATUS_APPROVED)
+            ->setTitle($productReview['title'])
+            ->setDetail($productReview['review'])
+            ->setEntityId($storeId)
+            ->setStores($storeId)
+            ->setCustomerId($this->getCustomerId($productReview['reviewer']['email']))
+            ->setNickname($productReview['reviewer']['name'])
+            ->save();
+
+        $this->saveReviewRatings($review, $productReview, $config);
+
+        $review->aggregate();
+
+        $this->logger->debug(sprintf('Saved product review with ID (%d)', $review->getId()));
+        return $review->getId();
+    }
+
+    private function isAuthorized(array $config, int $webshop_id, string $api_key): void {
+        if ($config['webshop_id'] == $webshop_id && $config['api_key'] == $api_key) {
```

Comment:
When comparing secrets you should use `hash_equals` which always takes the same amount of time to compare the string, instead of stopping at the first unequal character.

Otherwise in theory you could more quickly brute force the key by analyzing the amount of time it takes to generate the response (the more characters are equal, the longer the response will take, so this gives an attacker useful feedback).

https://en.wikipedia.org/wiki/Timing_attack
https://www.php.net/hash_equals